### PR TITLE
[SPARK-40625] Add MASK_CCN and TRY_MASK_CCN functions to redact credit card string values

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

Add MASK_CCN and TRY_MASK_CCN functions to redact credit card string values.

Each of these functions converts a string 'input' representing a credit card number to an updated version applying a transformation to the characters. This can be useful for creating copies of tables with sensitive information removed, but retaining the same schema.

Both functions return an error if the format string is invalid.

MASK_CCN returns an error if the input string does not match the format string.

TRY_MASK_CCN instead returns NULL in that case.

The format can consist of the following characters, case insensitive:
  - Each 'X' represents a digit which will be converted to 'X' in the result.
  - Each digit '0'-'9' represents a digit which will be left unchanged in the result.
  - Each '-' character should match exactly in the input string.
  - The default format string is: XXXX-XXXX-XXXX-XXXX.

No other format characters are allowed.
Any leading or trailing whitespace in the input string is stripped out.

Examples:
```
> SELECT MASK_CCN(ccn) FROM VALUES ("1234-5678-9876-5432") AS tab(ccn);
  XXXX-XXXX-XXXX-XXXX
> SELECT MASK_CCN("  1234-5678-9876-5432  ", "XXXX-XXXX-XXXX-1234");
  XXXX-XXXX-XXXX-5432
> SELECT MASK_CCN("[1234-5678-9876-5432]", "[XXXX-XXXX-XXXX-1234]");
  Error: the format string is invalid
> SELECT MASK_CCN("1234567898765432");
  Error: the input string does not match the format
> SELECT MASK_CCN("1234567898765432", "XXXX-XXXX-XXXX-1234");
  Error: the input string does not match the format
> SELECT TRY_MASK_CCN("1234567898765432");
  NULL
> SELECT TRY_MASK_CCN("1234567898765432", "XXXX-XXXX-XXXX-1234");
  NULL
```

### Why are the changes needed?

These functions are useful for processing string values.

### Does this PR introduce _any_ user-facing change?

Yes, it adds new SQL functions.

### How was this patch tested?

This PR adds a new unit test suite.

### CODE REVIEW

**Comment by dtenedor** [comment]
Hi @vitaliili-db can you please help review this when you have time? 🙏

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    _FUNC_(input) - Convert string 'input' representing a credit card number to an updated version
+    applying a transformation to the characters. This can be useful for creating copies of tables
+    with sensitive information removed, but retaining the same schema. Returns an error if the
+    format string is invalid or if the input string does not match the format string.
+    The format can consist of the following characters, case insensitive:
+      - Each 'X' represents a digit which will be converted to 'X' in the result.
+      - Each digit '0'-'9' represents a digit which will be left unchanged in the result.
+      - Each '-' character should match exactly in the input string.
+    No other format characters are allowed. Any leading or trailing whitespace in the input string
+    is stripped out. The default is: XXXX-XXXX-XXXX-XXXX.
```
Comment (vitaliili-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:45]: in code whitespaces are ignored.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+      > SELECT _FUNC_("  1234-5678-9876-5432  ", "XXXX-XXXX-XXXX-1234");
+        XXXX-XXXX-XXXX-5432
+      > SELECT _FUNC_("[1234-5678-9876-5432]", "[XXXX-XXXX-XXXX-1234]");
+        Error: the format string is invalid
+      > SELECT _FUNC_("1234567898765432");
+        Error: the input string does not match the format
+      > SELECT _FUNC_("1234567898765432", "XXXX-XXXX-XXXX-1234");
+        Error: the input string does not match the format
+  """,
+  since = "3.3.0",
```
Comment (vitaliili-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:60]: should be `3.4.0`?

Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:60]: +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        |boolean ${ev.isNull} = ${eval.isNull};
+        |${CodeGenerator.javaType(dataType)} ${ev.value} = ${CodeGenerator.defaultValue(dataType)};
+        |if (!${ev.isNull}) {
+        |  ${ev.value} = $builder.parse(${eval.value});
+        |  if (${ev.value} == null) {
+        |    ${ev.isNull} = true;
+        |  }
+        |}
+      """
+    val stripped = result.stripMargin
```
Comment (amaliujia) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:171]: nit: 

can be merged as 
```
"""
""".stripMargin
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+  def this(left: Expression) = this(left, Literal(MaskCcn.DEFAULT_FORMAT))
+  override protected def withNewChildrenInternal(
+      newInput: Expression, newFormat: Expression): TryMaskCcn =
+    copy(left = newInput, right = newFormat)
+  override def nullable: Boolean = true
+}
+
+/** Companion object for the MaskCcn and TryMaskCcn classes. */
+object MaskCcn {
+  val DEFAULT_FORMAT = "XXXX-XXXX-XXXX-XXXX"
```
Comment (amaliujia) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:116]: Just curious, will be `XXXX XXXX XXXX XXXX` be another common input format (replace `-` by whitespace)?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 5]
```diff
@@ -3147,4 +3147,18 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
...
   }
+
+  def maskCcnInvalidFormatError(functionName: String): Throwable = {
+    new AnalysisException(
+      errorClass = "REDACTION_FUNCTION_ERROR",
+      errorSubClass = "MASK_CCN_INVALID_FORMAT_ERROR",
+      messageParameters = Map("functionName" -> functionName))
+  }
+
+  def maskCcnFormatMatchError(input: String, format: String): Throwable = {
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:3158]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+/** Implementation of the 'mask_ccn' and 'try_mask_ccn' functions. */
+abstract class MaskCcnBase(
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:120]: Good idea, done!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+/** Executes the string parsing steps for the 'mask_ccn' and 'try_mask_ccn' functions. */
+class MaskCcnParser(formatString: String, nullOnError: Boolean) extends Serializable {
+  def parse(input: UTF8String): UTF8String = {
+    val inputString = input.toString
+    var formatStringIndex = 0
+    var error = false
+    // Check and consume each character in the input string, comparing against characters in the
+    // format string.
+    val result = inputString.map { inputChar =>
+      if (inputChar.isWhitespace) {
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:185]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala (hunk 2 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:60]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala (hunk 3 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:171]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala (hunk 4 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:116]: That's a good example! I updated the logic to preserve whitespaces in the input string, and skip/ignore whitespaces in the format string. This supports cases you mentioned, and might be flexible for other use cases later as well.

**Comment by dtenedor** [comment]
@vitaliili-db @amaliujia thanks so much for your review! I responded to your comments, please take another look.

**Review by vitaliili-db — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/RedactionFunctionsSuite.scala [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.{SparkFunSuite, SparkRuntimeException}
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+
+class RedactionFunctionsSuite extends SparkFunSuite with ExpressionEvalHelper {
```
Comment (amaliujia) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/RedactionFunctionsSuite.scala:24]: Am I missing anything, `MASK_CCN_INVALID_FORMAT_ERROR` is not tested yet?

**Review by amaliujia — COMMENTED** [review]
LGTM with one comment

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/RedactionFunctionsSuite.scala (hunk 8 above)
Comment (dtenedor) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/RedactionFunctionsSuite.scala:24]: It's covered in the `QueryCompilationErrors.maskCcnInvalidFormatError` method:

image

Comment (amaliujia) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/RedactionFunctionsSuite.scala:24]: oops yes thanks!

**Comment by dtenedor** [comment]
The only test failure is unrelated and flaky

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+      }
+    } else {
+      inputTypeCheck
+    }
+  }
+
+  override def nullSafeEval(string: Any, format: Any): Any =
+    parser.parse(string.asInstanceOf[UTF8String])
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:150]: How about overriding `nullSafeCodeGen` instead?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+/** Companion object for the MaskCcn and TryMaskCcn classes. */
+object MaskCcn {
+  val DEFAULT_FORMAT = "XXXX-XXXX-XXXX-XXXX"
+}
+
+/** Implementation of an expression to mask digits in a string to replacement characters. */
+abstract class MaskDigitSequence(
+    left: Expression, right: Expression, functionName: String, nullOnError: Boolean)
+  extends BinaryExpression with ImplicitCastInputTypes with NullIntolerant with Serializable {
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:116]: why extending Serializable

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+        |  if (${ev.value} == null) {
+        |    ${ev.isNull} = true;
+        |  }
+        |}
+      """.stripMargin)
+  }
+}
+
+/** Executes the string parsing steps for the MaskDigitSequence class. */
+class MaskDigitSequenceParser(formatString: String, nullOnError: Boolean) extends Serializable {
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:170]: why extending Serializable

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+    // format string.
+    val result = inputString.map { inputChar =>
+      if (error) {
+        // If we have encountered an error, leave the input character alone; we will raise an
+        // exception or return NULL after this loop has finished.
+        inputChar
+      } else if (inputChar.isWhitespace) {
+        // The input character is whitespace. Ignore it and continue comparing the next input
+        // character against the same format string character as this iteration.
+        inputChar
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:190]: Shouldn't we check if the corresponding format string char is space as well?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+      } else if (formatStringIndex >= formatString.length) {
+        // We have already consumed all the characters in the format string, but one or more
+        // characters exist in the input string. This is an error because the input string does not
+        // match the format string.
+        error = true
+        'X'
+      } else {
+        // Check the corresponding character in the format string.
+        skipFormatWhitespace()
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:199]: Seems we are going to skip the spaces on both input/format string...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+/** Implementation of an expression to mask digits in a string to replacement characters. */
+abstract class MaskDigitSequence(
+    left: Expression, right: Expression, functionName: String, nullOnError: Boolean)
+  extends BinaryExpression with ImplicitCastInputTypes with NullIntolerant with Serializable {
+  private def format: Expression = right
+
+  private lazy val formatString = if (format.foldable) {
+    format.eval().toString.toUpperCase(Locale.ROOT)
+  } else {
+    ""
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:122]: Then the format is invalid. We store an empty string here in order to check it below. I updated the `checkInputDataTypes` method accordingly and added some test coverage.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala (hunk 9 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:150]: Good idea, done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+  }
+}
+
+/** Executes the string parsing steps for the MaskDigitSequence class. */
+class MaskDigitSequenceParser(formatString: String, nullOnError: Boolean) extends Serializable {
+  def parse(input: UTF8String): UTF8String = {
+    val inputString = input.toString
+    var formatStringIndex = 0
+    var error = false
+    def skipFormatWhitespace(): Unit =
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:175]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala (hunk 12 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:190]: Good question, I think the intention is to skip and ignore whitespace in both the input string and the format string. For example, for use cases like credit card numbers, phone numbers, and social security numbers, these are not material to the data in the field. I put a comment to this effect here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala (hunk 13 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:199]: Same, this is per intention (left a comment).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:122]: How about throwing exception in `checkInputDataTypes` if the format is not foldable? The error will be much more clear.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala (hunk 11 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:170]: This one was not necessary, I have removed it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala (hunk 10 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:116]: This is needed, it makes the subclass Expression serializable. Removing it causes the tests to fail with serialization exceptions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala:122]: Good idea, done!

**Comment by dtenedor** [comment]
@gengliangwang thanks for review! 👍  Responded to your comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala (hunk 12 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala]: Update on this: per our offline conversation, we decided to require whitespace in the format string to match the input string exactly. I updated the PR accordingly.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala (hunk 13 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala]: Update on this: per our offline conversation, we decided to require whitespace in the format string to match the input string exactly. I updated the PR accordingly.

**Comment by dtenedor** [comment]
@gengliangwang @vinodkc FYI, it looks like there is a duplication of planned effort between this PR and https://issues.apache.org/jira/browse/SPARK-40686. We should probably dedup these into one effort. @vinodkc do you want to take this on, and I can close my Jira and help with the review?

**Comment by dtenedor** [comment]
going to close this in favor of the other work by @vinodkc instead

**Pull request closed without merging**